src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,185 @@
+        rightSpeedController = new SpeedControllerGroup(frontRightSpark, rearRightSpark);
+
+        frontLeftEncoder = frontLeftSpark.getEncoder();
+        frontRightEncoder = frontRightSpark.getEncoder();
+        rearLeftEncoder = rearLeftSpark.getEncoder();
+        rearRightEncoder = rearRightSpark.getEncoder();
+
+        robotDrive = new DifferentialDrive(leftSpeedController, rightSpeedController);
+
+        frontLeftPositionController = new SparkPositionController(frontLeftSpark, smartMotionSettings);
+        frontRightPositionController = new SparkPositionController(frontRightSpark, smartMotionSettings);
+        rearLeftPositionController = new SparkPositionController(rearLeftSpark, smartMotionSettings);
+        rearRightPositionController = new SparkPositionController(rearRightSpark, smartMotionSettings);
+        
+        posController = new SparkPositionControllerGroup(frontLeftPositionController, frontRightPositionController,
+                rearLeftPositionController, rearRightPositionController);
+        autoController = new PositionDriveController(frontLeftSpark, frontRightSpark, 
+                rearLeftSpark, rearRightSpark, smartMotionSettings, positionBuffer,
+                new EncoderInchesConverter(1/ RobotConstants.DIMENSIONS.MOTOR_REVOLUTIONS_PER_INCH));
+        reset();
+    }
+
+    public void setSpeedMode() {
+        speedSettings.configureSparkMax(frontLeftSpark);
+        speedSettings.configureSparkMax(frontRightSpark);
+        speedSettings.configureSparkMax(rearLeftSpark);
+        speedSettings.configureSparkMax(rearRightSpark);
+    }
+
+    public void setPositionMode() {
+        positionSettings.configureSparkMax(frontLeftSpark);
+        positionSettings.configureSparkMax(frontRightSpark);
+        positionSettings.configureSparkMax(rearLeftSpark);
+        positionSettings.configureSparkMax(rearRightSpark);
+    }
+
+    public EncoderValues getEncoderValues() {
+        return new EncoderValues(frontLeftEncoder.getPosition(),
+                rearLeftEncoder.getPosition(),
+                frontRightEncoder.getPosition(),
+                rearRightEncoder.getPosition());
+    }
+
+    @Override
+    public void customPeriodic(RobotPose rp, FieldPose fp) {
+        if (inAuto) {
+            autoController.periodic();
+        }
+        logger.log("Front Left Encoder Ticks", frontLeftEncoder.getPosition());
+        logger.log("Front Right Encoder Ticks", frontRightEncoder.getPosition());
+        logger.log("Rear Left Encoder Ticks", rearLeftEncoder.getPosition());
+        logger.log("Rear Right Encoder Ticks", rearRightEncoder.getPosition());
+        if (getCurrentCommand() != null) {
+            logger.log("current command", getCurrentCommand().getName());
+        }
+    }
+
+    public void drive(DriveInstruction di) {
+        robotDrive.arcadeDrive(di.getFoward(), di.getRotation());
+    }
+
+    public void startAutonomous() {
```

Comment:
i'd rather see this in Robot. there's already a method called autonomousInit(). why do we need another one here that says start autonomous.? if we need to do things on mode switch, we have setSpeedMode and setPositionMode above